### ISSUE TITLE
Euler Angles to Rotation Matrix Conversion Inconsistent with Scipy

### ISSUE THREAD

**Issue by lordfjw** [facebookresearch/pytorch3d#649]
## 🐛 Bugs / Unexpected behaviors

Euler Angles to Rotation Matrix Conversion with pytorch3d has inconsistent results with scipy.

## Instructions To Reproduce the Issue:

1. Code to Reproduce
```
import torch
import functools
from scipy.spatial.transform import Rotation
import numpy as np
from copy import deepcopy

def _axis_angle_rotation(axis: str, angle):
  """
  Return the rotation matrices for one of the rotations about an axis
  of which Euler angles describe, for each value of the angle given.
  Args:
      axis: Axis label "X" or "Y or "Z".
      angle: any shape tensor of Euler angles in radians
  Returns:
      Rotation matrices as tensor of shape (..., 3, 3).
  """

  cos = torch.cos(angle)
  sin = torch.sin(angle)
  one = torch.ones_like(angle)
  zero = torch.zeros_like(angle)

  if axis == "X":
    R_flat = (one, zero, zero, zero, cos, -sin, zero, sin, cos)
  if axis == "Y":
    R_flat = (cos, zero, sin, zero, one, zero, -sin, zero, cos)
  if axis == "Z":
    R_flat = (cos, -sin, zero, sin, cos, zero, zero, zero, one)

  return torch.stack(R_flat, -1).reshape(angle.shape + (3, 3))


def euler_angles_to_matrix(euler_angles, convention: str):
  """
  Convert rotations given as Euler angles in radians to rotation matrices.
  Args:
      euler_angles: Euler angles in radians as tensor of shape (..., 3).
      convention: Convention string of three uppercase letters from
          {"X", "Y", and "Z"}.
  Returns:
      Rotation matrices as tensor of shape (..., 3, 3).
  """
  if euler_angles.dim() == 0 or euler_angles.shape[-1] != 3:
    raise ValueError("Invalid input euler angles.")
  if len(convention) != 3:
    raise ValueError("Convention must have 3 letters.")
  if convention[1] in (convention[0], convention[2]):
    raise ValueError(f"Invalid convention {convention}.")
  for letter in convention:
    if letter not in ("X", "Y", "Z"):
      raise ValueError(f"Invalid letter {letter} in convention string.")
  matrices = map(_axis_angle_rotation, convention, torch.unbind(euler_angles, -1))
  # return matrices for testing
  return matrices
  # return functools.reduce(torch.matmul, matrices)


def test_conversion():
  euler_angles = np.array([0.3, -0.5, 0.7], dtype=np.float32)
  rot_mat_scipy = Rotation.from_euler('xyz', euler_angles, degrees=False).as_matrix()
  euler_angles_torch = torch.from_numpy(euler_angles.astype(np.float32)).unsqueeze(0)
  matrices = euler_angles_to_matrix(euler_angles_torch, 'XYZ')
  matrices_list = list(deepcopy(matrices))
  rot_mat_torch3d = functools.reduce(torch.matmul, matrices)
  rot_mat_xyz = torch.matmul(torch.matmul(matrices_list[0], matrices_list[1]), matrices_list[2])
  rot_mat_zyx = torch.matmul(torch.matmul(matrices_list[2], matrices_list[1]), matrices_list[0])
  print('Input Euler anlges radians in XYZ: \n {}'.format(euler_angles))
  print('Rotation Matrix from scipy: \n {}'.format(rot_mat_scipy))
  print('Rotation Matrix from pytorch3d: \n {}'.format(rot_mat_torch3d))
  print('Rotation Matrix from pytorch3d, multiplying as Rx*Ry*Rz: \n {}'.format(rot_mat_xyz))
  print('Rotation Matrix from pytorch3d, multiplying as Rz*Ry*Rx: \n {}'.format(rot_mat_zyx))


if __name__ == '__main__':
  test_conversion()
```
2. The exact command(s) you ran: simply run the script provided
3. What you observed (including the full logs):
```
Input Euler anlges radians in XYZ: 
 [ 0.3 -0.5  0.7]
Rotation Matrix from scipy: 
 [[ 0.67121217 -0.72380745 -0.1599281 ]
 [ 0.5653542   0.63940893 -0.52108622]
 [ 0.47942554  0.25934339  0.83838664]]
Rotation Matrix from pytorch3d: 
 tensor([[[ 0.6712, -0.5654, -0.4794],
         [ 0.5071,  0.8220, -0.2593],
         [ 0.5407, -0.0690,  0.8384]]])
Rotation Matrix from pytorch3d, multiplying as Rx*Ry*Rz: 
 tensor([[[ 0.6712, -0.5654, -0.4794],
         [ 0.5071,  0.8220, -0.2593],
         [ 0.5407, -0.0690,  0.8384]]])
Rotation Matrix from pytorch3d, multiplying as Rz*Ry*Rx: 
 tensor([[[ 0.6712, -0.7238, -0.1599],
         [ 0.5654,  0.6394, -0.5211],
         [ 0.4794,  0.2593,  0.8384]]])

```

## Comment
From the code, it's clear that pytorch3d first compute 3 rotation matrices Rx, Ry and Rz with _axis_angle_rotation, and simply multiply with Rx*Ry*Rz. You can confirm that from the output that the Rotation Matrix from pytorch3d is exactly the same with the one from Rx*Ry*Rz. Geometrically, this means first apply the rotation around z axis, then apply rotation around y axis, and then apply rotation around x axis, namely composition of extrinsic rotation. However, this is not consistent with scipy rotation conversion, which uses composition of intrinsic rotation, namely getting rotation matrix R=Rz*Ry*Rx. 

I am not sure whether the extrinsic rotation convention here is intentional or not. If it is, then we should at least mention it in the document to avoid further confusion. Thanks!

For reference:
https://en.wikipedia.org/wiki/Rotation_matrix
https://en.wikipedia.org/wiki/Euler_angles#Intrinsic_rotations

**Comment by nikhilaravi** [comment]
@lordfjw thanks for explaining this. We will look into it and get back to you.

**Comment by bottler** [comment]
PyTorch3D is using extrinsic rotations. I agree that we could mention this.
Scipy gives a choice between extrinsic (with capital letters) and intrinsic rotations (with small letters). I don't agree that scipy prefers intrinsic ones, or that we are inconsistent with it. The following three commands give the same results
```
euler_angles_to_matrix(torch.tensor(euler_angles), 'XYZ')
Rotation.from_euler('XYZ', euler_angles, degrees=False).as_matrix()
Rotation.from_euler('zyx', euler_angles[::-1], degrees=False).as_matrix()
```

**Comment by caizhongang** [comment]
Hi, from Scipy [documentation ](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_euler.html)

> ... the set {‘X’, ‘Y’, ‘Z’} for intrinsic rotations, or {‘x’, ‘y’, ‘z’} for extrinsic rotations


Consider the expriments below:

> ```
> euler_angles_to_matrix(torch.tensor(euler_angles), 'XYZ')
> Rotation.from_euler('XYZ', euler_angles, degrees=False).as_matrix()
> Rotation.from_euler('zyx', euler_angles[::-1], degrees=False).as_matrix()
> ```

and the original comment:

> From the code, it's clear that pytorch3d first compute 3 rotation matrices Rx, Ry and Rz with _axis_angle_rotation, and simply multiply with RxRyRz. You can confirm that from the output that the Rotation Matrix from pytorch3d is exactly the same with the one from RxRyRz. Geometrically, this means first apply the rotation around z axis, then apply rotation around y axis, and then apply rotation around x axis, namely composition of extrinsic rotation. However, this is not consistent with scipy rotation conversion, which uses composition of intrinsic rotation, namely getting rotation matrix R=RzRyRx.

Is Pytorch3D actually using **intrinsic** rotation?

**Comment by bottler** [comment]
Yes, looks like the words extrinsic and intrinsic are switched in my previous comment, and possibly also in the OP.

I won't fix it above because I think that would make this thread more confusing.
